src/block/Campfire.php

```diff
@@ -0,0 +1,193 @@
+use pocketmine\player\Player;
+use pocketmine\world\BlockTransaction;
+use pocketmine\world\sound\FireExtinguishSound;
+use pocketmine\world\sound\FlintSteelSound;
+use pocketmine\world\sound\ItemFrameAddItemSound;
+use function count;
+
+class Campfire extends Transparent{
+	use HorizontalFacingTrait;
+
+	protected bool $extinguished = false;
+
+	public function writeStateToMeta() : int{
+		return $this->facing | ($this->extinguished ? BlockLegacyMetadata::CAMPFIRE_FLAG_EXTINGUISHED : 0);
+	}
+
+	public function readStateFromData(int $id, int $stateMeta) : void{
+		$this->facing = $stateMeta & 0x03;
+		$this->extinguished = ($stateMeta & BlockLegacyMetadata::CAMPFIRE_FLAG_EXTINGUISHED) === BlockLegacyMetadata::CAMPFIRE_FLAG_EXTINGUISHED;
+	}
+
+	public function getStateBitmask() : int{
+		return 0b111;
+	}
+
+	public function hasEntityCollision() : bool{
+		return true;
+	}
+
+	public function getLightLevel() : int{
+		return $this->extinguished ? 0 : 15;
+	}
+
+	public function isAffectedBySilkTouch() : bool{
+		return true;
+	}
+
+	public function getDropsForCompatibleTool(Item $item) : array{
+		return [
+			VanillaItems::CHARCOAL()->setCount(2)
+		];
+	}
+
+	public function isExtinguished() : bool{
+		return $this->extinguished;
+	}
+
+	/** @return $this */
+	public function setExtinguished(bool $extinguish = true) : self{
+		$this->extinguished = $extinguish;
+		return $this;
+	}
+
+	private function extinguish() : void{
+		$this->position->getWorld()->addSound($this->position, new FireExtinguishSound());
+		$this->position->getWorld()->setBlock($this->position, $this->setExtinguished());
+	}
+
+	private function fire() : void{
+		$this->position->getWorld()->addSound($this->position, new FlintSteelSound());
+		$this->position->getWorld()->setBlock($this->position, $this->setExtinguished(false));
+		$this->position->getWorld()->scheduleDelayedBlockUpdate($this->position, 1);
+	}
+
+	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
+		if($this->getSide(Facing::DOWN)->isTransparent()){
+			return false;
+		}
+		if($player !== null){
+			$faces = [
+				4 => 3,
+				3 => 2,
+				2 => 0,
+				5 => 1
+			];
+			$this->facing = $faces[$player->getHorizontalFacing()];
+		}
+		return parent::place($tx, $item, $blockReplace, $blockClicked, $face, $clickVector, $player);
+	}
+
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
+		if($player !== null){
+			if($item instanceof FlintSteel){
+				if($this->extinguished){
+					$item->applyDamage(1);
+					$this->fire();
+				}
+				return true;
+			}
+			if($item instanceof Shovel && !$this->extinguished){
+				$item->applyDamage(1);
+				$this->extinguish();
+				return true;
+			}
+
+			$tile = $this->position->getWorld()->getTile($this->position);
+			if($tile instanceof TileCampfire && $tile->addItem(clone $item)){
```

Comment:
You should not interact with the tile except in `readStateFromWorld` and `writeStateFromWorld`. See, for example, `ItemFrame`.

Reply:
There should also be APIs on the Block class to change the items in the campfire without touching the tile.